tutorials/14-minituring/14_minituring.jmd

```diff
@@ -0,0 +1,300 @@
+    
+    # Data
+    args
+    data # a tuple of all the data
+end
+```
+
+To illustrate how inference work for our mini language. We will implement an extremely simplistic Random-Walk Metropolis-Hastings sampler.
+
+> **Julia Sidebar**  
+We used the Julia package [AbstractMCMC.jl](https://github.com/TuringLang/AbstractMCMC.jl) here. For the interface design, we refer the reader to the [documentation](https://beta.turing.ml/AbstractMCMC.jl/dev/).  
+In simple words, we need to implement a sub-type of `AbstractModel`, which we did in the code block above; a sub-type of `AbstractSampler`, which we'll do in the next code block; and `step` functions, following the sampler definition.
+
+```julia
+struct MHSampler <: AbstractMCMC.AbstractSampler
+    proposal # Just a Normal distribution in this simple case
+end
+
+abstract type AbstractTransition end
+struct Transition <: AbstractTransition
+    varinfo :: AbstractVarInfo
+end
+```
+
+We hard-coded the proposal step as part of a `context` implementation. In Turing, to accommodate more use-cases, the `proposal` function is abstracted out.
+
+```julia
+struct MHTransitionContext <: AbstractContext 
+    rng::Random.AbstractRNG
+end
+
+MHTransitionContext() = MHTransitionContext(_GLOBALRNG_)
+
+function _observe(__context__::MHTransitionContext, __varinfo__, sampler, dist, var_id, var_value) 
+    log_p = Distributions.loglikelihood(dist, var_value)
+    update_varinfo!(__varinfo__, var_id, var_value, log_p)
+end
+
+function _assume(__context__::MHTransitionContext, __varinfo__, sampler, dist, var_id)
+    old_value = __varinfo__.values[var_id]
+    # propose a random-walk step, i.e, add the current value to a random value sampled from
+    # a Normal distribution centered at 0
+    new_value = old_value + rand(__context__.rng, sampler.proposal)
+    log_p = Distributions.loglikelihood(dist, new_value)
+    update_varinfo!(__varinfo__, var_id, new_value, log_p)
+    return new_value
+end
+```
+
+We need to define two `step` functions, one for the first step and the other for the following steps. The two functions are identified with different arguments they take.  
+
+```julia
+# The fist step
+function AbstractMCMC.step(
+    rng::Random.AbstractRNG,
+    model::MiniModel,
+    sampler::MHSampler;
+    kwargs...
+)
+    vi = VarInfo()
+    ctx = SamplingContext()
+    model.f(model, vi, ctx, model.data...)
+    return Transition(vi), vi
+end
+
+# Defining functions to compute values to determine if accept the proposal
+function logα(old_vi, new_vi, args, proposal)
+    return compute_joint(new_vi) - compute_joint(old_vi) 
+    + compute_condition_log(old_vi, new_vi, args, proposal) - 
+    compute_condition_log(new_vi, old_vi, args, proposal)
+end
+    
+function compute_condition_logprob(vi, condition_vi, args, proposal)
+    return reduce(+, [Distributions.loglikelihood(proposal, value1 - value2) 
+            for ((key, value1), (_, value2)) in zip(vi.values, condition_vi.values) if key ∉ args])
```

Comment:
This is a bit complicated. It also seems possibly wrong since it is not guaranteed that the keys of `value1` and `value2` are the same since `Dict` is not an ordered dictionary.
An initial simplification would be e.g.
```suggestion
    values1 = vi.values
    values2 = condition_vi.values
    return sum([loglikelihood(proposal, values1[key] - value2[key]) for key in keys(values1) if !(key in args)])
```